src/execution.rs

```diff
@@ -59,6 +59,9 @@ impl File {
         globals: &mut Variables,
     ) -> Result<Graph<'tree>, ExecutionError> {
         let mut graph = Graph::new();
+        if tree.root_node().has_error() {
+            return Err(ExecutionError::ParseTreeHasErrors);
+        }
```

Comment:
Do I understand correctly that this means that if a file has but a single syntax error, it will get ignored completely?

From my perspective (and use case) that seems rather severe. I would rather continue with the execution regardless (since it may very well be that the syntax error has a limited impact on the rest of the file).

At the very least, we could make this behaviour dependent on a command-line flag.

Reply:
TLDR: If this check is a problem for you right now, I'll take it out. I may have an idea how to do this more fine-grained, but that'll be a different PR.

> Do I understand correctly that this means that if a file has but a single syntax error, it will get ignored completely?

Yes.

> From my perspective (and use case) that seems rather severe. I would rather continue with the execution regardless

I agree that this would be more desirable. The reason I did it like this (for now) is because of similar problems we had with `tree-sitter-code-nav`. There executing stanzas on trees with errors would sometimes result in hard panics---also not ideal for a library.

I haven't had time to investigate exactly if we can be more precise while still avoiding the errors. Let's assume that query matching is somewhat sane, so a query only matches if all named nodes in the pattern are actually there in the tree. That means we can only refer to error nodes in leaf, wildcard captures. If the capture is an error node itself, using the capture will probably always result in an error:
1. Accessing the source text might be a problem. I think this was the main reason for the panics in `tree-sitter-code-nav`, when there were parse errors for identifier names.
2. It is (probably) missing expected scoped variables defined on it. The capture could also have error nodes deeper in the subtree, which may or may not prevent other stanzas to fire, which may also result in errors because of missing scoped variables.

Okay, now that I am writing this up, one strategy that may work is introducing a special exception that indicates we hit an error node during evaluation. We maybe only have to throw it when we convert a syntax reference to the actual node---thereby guaranteeing that if we get a node, it is never an error node (although it may have error nodes as descendants). Instead of always propagating this exception to the top, we can absorb it at a statement level. So, any statement that involves an error node will be skipped. However, this this not solve the missing scoped variables completely, because values may not be set if they hit an error node. If we can also mark (scoped) variables as error-like if their value could not be computed because of an error node, then we have enough to avoid missing (scoped) variable errors, I think.

I think this is worth trying, but I'll do that in a different PR. If this check is problematic for you atm, I'll take it out. Let me know.

Reply:
Hmm... You're right about things likely failing due to missing scoped variables.

In that case, I think it's probably better to include this change.

Truth be told, I'm not entirely sure how `ERROR` nodes work. I had assumed that they were just a slightly odd kind of named node (with a location and potentially with children). However, seeing as it isn't possible to query for these nodes, I'm not too sure how accurate this mental image is.

Ideally, I would want to be able to query for these `ERROR` nodes and also capture and assign scoped variables to them. That way, I can just treat them as another kind of AST node, link them up appropriately (without running into missing scoped variables), and then recognise and handle them appropriately in later stages of the analysis.